### DISCUSSION TITLE
Context-Free Inline Assists

### DISCUSSION THREAD

**Discussion by NiloCK** [zed-industries/zed#23253] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

I'm a heavy user of the assistant panel. I often have long active chats. But the bundling of these chats adds cost overhead to small, inline, and unrelated assistance requests.

The **most common** examples are when I ask about specific commands inside the terminal: they rarely require any outside context, but are substantially more expensive than if I run them with a blank context.

### Environment

Zed: v0.164.2 (Zed)
OS: Linux Wayland ubuntu 22.04
Memory: 15.4 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics (CML GT2) || Intel open-source Mesa driver || Mesa 23.2.1-1ubuntu3.1~22.04.2

### If applicable, add mockups / screenshots to help present your vision of the feature

One possible option could be to add a 'clear' or 'trash' icon next to the context notifier during an inline-assist prompt.

![image](https://github.com/user-attachments/assets/5c806a41-fda9-46f8-a643-75477e269003)

Another could be to make a separate command to launch a fresh-context inline assist. Eg, `ctrl-shift-enter` instead of `ctrl-enter`.

**Comment by anzax** [comment]
Using inline assist with full context is unique and super powerful feature, but I agree with @NiloCK, often there is a need to do small independent edit so it would be nice to have a toggle or/and separate shortcut to do edit without full context (shift+enter maybe). 
@NiloCK, simple workaround is to create new chat (cmd+?, cmd+n, cmd+?), I often have multiple active chats and switch between them for different local edits.

**Comment by notpeter** [comment]
There are significant changes to how context is shared between the inline assistant and assistant panel with the forthcoming Assistant2.  As such, this enhancement request is unlikely to be implemented as requested and so I'm going to move this to a discussion.  Thanks for reporting!

**Reply by anzax** [reply]
Hi @notpeter, could you please share a link to the issue or documentation where we can read about how Assistant2 handles context? I find the current implementation both unique and powerful, but it does require a good level of awareness and knowledge to use effectively. I hope the ability to leverage chat context for local edits remains intact, as it’s incredibly useful.

**Reply by yangchenyun** [reply]
Is there a place to look up for the scope / improvement of Assistant2 and when it would be ready to use?

**Reply by notpeter** [reply]
It's very much under active development and so many things (including scope and timeline) are still TBD, but initial release is I think on-track for mid-Q1.  One of the core reasons behind implementing the rewrite is to better support different methods of context inclusion in addition to the existing manually adding/pasting/editing context.  The current Assistant is a largely implemented as a monolithic context focused on a single model which Inline Assistant is forced to share.

With Assistant2 (work-in-progress, no promises, not seeking feedback). You can explicitly control the context for an Inline Assistant, so you can include/remove files from context, add one (or multiple) Assistant Panel Chat Conversations, etc.
<img width="928" alt="Screenshot 2025-01-17 at 12 24 46" src="https://github.com/user-attachments/assets/929fbd7b-c83a-42d9-aed8-8f46b8983714" />

Things are actively being developed and significant iterative design works remain, but the spirit of this enhancement request (control context for inline assistant) is very much reflected in the direction we're looking to head. Assistant2 is not intended to deprecate/immediately replace the existing monolithic assistant, but explore additional new methods of interaction.

**Reply by 2nthony** [reply]
> With Assistant2 (work-in-progress, no promises, not seeking feedback). You can explicitly control the context for an Inline Assistant, so you can include/remove files from context, add one (or multiple) Assistant Panel Chat Conversations, etc. <img alt="Screenshot 2025-01-17 at 12 24 46" width="928" src="https://private-user-images.githubusercontent.com/145113/404390381-929fbd7b-c83a-42d9-aed8-8f46b8983714.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU2MDIxODYsIm5iZiI6MTc0NTYwMTg4NiwicGF0aCI6Ii8xNDUxMTMvNDA0MzkwMzgxLTkyOWZiZDdiLWM4M2EtNDJkOS1hZWQ4LThmNDZiODk4MzcxNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNVQxNzI0NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YmJlMWM2NzNkYjcxM2RhOGVjYmMwYzVmODliNzQ4NWIzNWFkNjE2OGVhYjIwNmRlNGE0MjYwZjBhMWYyN2ZiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.SlPmvm_nuK0q4SejxSYDynGgW7zu06WwWUz49B4-U7E">

seems like cursor, would like to know when will release

**Reply by notpeter** [reply]
It's available in limited Beta now. Signup for the waitlist here:
- https://zed.dev/ai/agent

Release is a couple weeks away.
